students/JackieCao/Lesson05/mailroom3.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python3
+"""
+mailroom3
+Improve mailroom by adding exceptions and comprehensions.
+"""
+donors = {"Fred Jones":[100, 200, 300],
+        "Amy Shumer": [2000, 4000, 1000],
+        "Bean Shing": [20, 400],
+        "Ann Shaw": [2, 400],
+        "King May": [200, 400, 100]
+        }
+
+# print donor names
+def print_donors():
+    for i in donors:
+        print(i)
+
+
+def thank_you():
+    # ask user for donor's name
+    while True:
+        # add exception
+        try:
+            name = input("Enter a donor's name "
+                        "(or 'list' to see all donors or 'menu' to exit)>")
+        except (KeyboardInterrupt, EOFError):
+            return None
+        if name == "list":
+            print_donors()
+        elif name == "menu":
+            return
+        else:
+            break
+
+    # ask user for donation amount
+    while True:
+        # add exception
+        try:
+            dollar = input("Enter a donation amount (or 'menu' to exit)>")
+        except (KeyboardInterrupt, EOFError):
+            return None
+        if dollar == "menu":
+            return
+        # add exception
+        try:
+            dollar = float(dollar)
+        except ValueError:
+            print("Input must be a number")
+        else:
+            break
+
+    # add new name and donation amount to the donors history
+    if name.title() in donors:
+        # use comprehensions
+        who = [i for i in donors if name.lower() == i.lower()]
+        donors[''.join(who)].append(dollar)
+    else:
+        l_dollar = [dollar]
+        donors[name] = l_dollar
+    # print the thank you letter
+    d = {'d_name': name, 'd_dollar': dollar}
+
+    print("Dear {d_name},\n\n    Thank you for your donation of ${d_dollar}.\n\nBest,\nDonation Group".format(**d))
+
+def thank_to_all():
+    for name in donors:
+        total_donated = sum(donors[name])
+        d = {'d_name': name, 'd_dollar': total_donated}
+        letter = "Dear {d_name},\n\n    Thank you for your donation of ${d_dollar}.\n\nBest,\nDonation Group".format(**d)
+        with open(name + '.txt', 'w') as f:
+            f.write(letter)
+        print("The thank you letter for {d_name} has been sent".format(**d))
+
+def make_report():
+    # use comprehensions
+    total_donated = [sum(donors[name]) for name in donors]
+    number_of_donations = [len(donors[name]) for name in donors]
+    average_donation = [t/d for t, d in zip(total_donated,
+                                            number_of_donations)]
+    names = [name for name in donors]
+    data = [(name, dollar, num, ave) for name, dollar, num, ave in zip(names,
+                                                                       total_donated,
+                                                                       number_of_donations,
+                                                                       average_donation)]
+    # sort the data
+    data.sort(key=lambda tup: tup[0])
```

Comment:
nice lambda!

but I think it would have sorted that way anyway :-) -- but did you want to sort by the name?